lib/resumption/handover.go

```diff
@@ -0,0 +1,265 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resumption
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"io"
+	"net"
+	"net/netip"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/multiplexer"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+const sockSuffix = ".sock"
+
+func sockPath(dataDir string, token resumptionToken) string {
+	hash := sha256.Sum256(token[:])
+	// unix domain sockets are limited to 108 or 104 characters, so the full
+	// sha256 hash is a bit too much (64 bytes in hex or 44 in b64); truncating
+	// the hash to 128 bits still gives us more than enough headroom to just
+	// assume that we'll have no collisions (a probability of one in a
+	// quintillion with 26 billion concurrent connections)
+	return filepath.Join(dataDir, "handover", base64.RawURLEncoding.EncodeToString(hash[:16])+sockSuffix)
+}
+
+func sockDir(dataDir string) string {
+	return filepath.Join(dataDir, "handover")
+}
+
+var errNoDataDir error = &trace.NotFoundError{Message: "data dir not configured"}
+
+func (r *SSHServerWrapper) attemptHandover(conn *multiplexer.Conn, token resumptionToken) {
+	handoverConn, err := r.dialHandover(token)
+	if err != nil {
+		if trace.IsNotFound(err) {
+			r.log.Debug("Resumable connection not found or already deleted.")
+			_, _ = conn.Write([]byte{notFoundServerExchangeTag})
+			return
+		}
+		r.log.WithError(err).Error("Error while connecting to handover socket.")
+		return
+	}
+	defer handoverConn.Close()
+
+	var remoteIP netip.Addr
+	if t, _ := conn.RemoteAddr().(*net.TCPAddr); t != nil {
+		remoteIP, _ = netip.AddrFromSlice(t.IP)
+	}
+	remoteIP16 := remoteIP.As16()
```

Comment:
I'm comfortable leaving the remote IP slice as all zeroes and let the server decide if that's good enough or not, rather than prohibit it in the client side.